File: ios/App/MyTasksWidget/TaskService.swift
import Foundation

/// Handles API communication with the MySDAManager backend for widget data.
/// Fetches tasks and supports completing tasks via the REST API.
class TaskService {

    static let baseURL = "https://mysdamanager.com/api/v1/tasks/widget"

    // MARK: - Fetch Tasks

    /// Fetches the user's pending tasks from the API.
    /// Falls back to cached data if the network request fails.
    /// - Parameter limit: Maximum number of tasks to return (default 10)
    /// - Returns: Result containing an array of TaskItems or an error
    static func fetchTasks(limit: Int = 10) async -> Result<[TaskItem], Error> {
        guard let token = SharedStorage.getSessionToken() else {
            return .failure(WidgetError.notAuthenticated)
        }

        guard let url = URL(string: "\(baseURL)?limit=\(limit)") else {
            return .failure(WidgetError.invalidURL)
        }

        var request = URLRequest(url: url)
        request.httpMethod = "GET"
        request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        request.setValue("application/json", forHTTPHeaderField: "Accept")
        request.setValue("MySDAManager-iOS-Widget/1.0", forHTTPHeaderField: "User-Agent")
        request.timeoutInterval = 15

        do {
            let (data, response) = try await URLSession.shared.data(for: request)

            guard let httpResponse = response as? HTTPURLResponse else {
                return fallbackToCacheOrFail(WidgetError.invalidResponse)
            }

            switch httpResponse.statusCode {
            case 200:
                let decoder = JSONDecoder()
                let widgetResponse = try decoder.decode(WidgetResponse.self, from: data)
                // Cache successful response for offline fallback
                SharedStorage.cacheTasks(widgetResponse.data)
                SharedStorage.updateCacheTimestamp()
                return .success(widgetResponse.data)

            case 401:
                // Try to refresh the token before giving up
                let refreshResult = await refreshToken()
                if refreshResult {
                    // Retry the request with the new token
                    return await fetchTasks(limit: limit)
                }
                SharedStorage.clearSessionToken()
                return .failure(WidgetError.sessionExpired)

            case 403:
                return .failure(WidgetError.forbidden)

            case 429:
                return fallbackToCacheOrFail(WidgetError.rateLimited)

            case 500...599:
                return fallbackToCacheOrFail(WidgetError.serverError(httpResponse.statusCode))

            default:
                return fallbackToCacheOrFail(
                    WidgetError.unexpectedStatus(httpResponse.statusCode)
                )
            }
        } catch let error as DecodingError {
            return fallbackToCacheOrFail(WidgetError.decodingFailed(error.localizedDescription))
        } catch {
            // Network error - try cache
            return fallbackToCacheOrFail(error)
        }
    }

    // MARK: - Complete Task

    /// Marks a task as completed via the API.
    /// - Parameter taskId: The ID of the task to complete
    /// - Returns: True if the task was successfully completed
    static func completeTask(taskId: String) async -> Bool {
        guard let token = SharedStorage.getSessionToken() else {
            return false
        }

        guard let url = URL(string: baseURL) else {
            return false
        }

        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue("MySDAManager-iOS-Widget/1.0", forHTTPHeaderField: "User-Agent")
        request.timeoutInterval = 10

        let body: [String: String] = ["taskId": taskId]
        request.httpBody = try? JSONEncoder().encode(body)

        do {
            let (_, response) = try await URLSession.shared.data(for: request)
            guard let httpResponse = response as? HTTPURLResponse else { return false }
            return httpResponse.statusCode == 200
        } catch {
            return false
        }
    }

    // MARK: - Token Refresh

    /// Attempts to refresh the session token using the stored refresh token.
    /// - Returns: True if the token was successfully refreshed
    private static func refreshToken() async -> Bool {
        guard let refreshToken = SharedStorage.getRefreshToken() else {
            return false
        }

        guard let url = URL(string: "https://mysdamanager.com/api/v1/auth/refresh") else {
            return false
        }

        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.timeoutInterval = 10

        let body: [String: String] = ["refreshToken": refreshToken]
        request.httpBody = try? JSONEncoder().encode(body)

        do {
            let (data, response) = try await URLSession.shared.data(for: request)
            guard let httpResponse = response as? HTTPURLResponse,
                  httpResponse.statusCode == 200 else {
                return false
            }

            struct RefreshResponse: Codable {
                let sessionToken: String
                let refreshToken: String?
            }

            let refreshResponse = try JSONDecoder().decode(RefreshResponse.self, from: data)
            SharedStorage.storeSessionToken(refreshResponse.sessionToken)
            if let newRefreshToken = refreshResponse.refreshToken {
                SharedStorage.storeRefreshToken(newRefreshToken)
            }
            return true
        } catch {
            return false
        }
    }

    // MARK: - Cache Fallback

    /// Attempts to return cached tasks when the network request fails.
    /// Returns the original error if no cache is available.
    private static func fallbackToCacheOrFail(_ error: Error) -> Result<[TaskItem], Error> {
        if let cachedTasks = SharedStorage.getCachedTasks(), !cachedTasks.isEmpty {
            return .success(cachedTasks)
        }
        return .failure(error)
    }
}

// MARK: - Widget Errors

enum WidgetError: LocalizedError {
    case notAuthenticated
    case sessionExpired
    case forbidden
    case invalidURL
    case invalidResponse
    case rateLimited
    case serverError(Int)
    case unexpectedStatus(Int)
    case decodingFailed(String)

    var errorDescription: String? {
        switch self {
        case .notAuthenticated:
            return "Not signed in"
        case .sessionExpired:
            return "Session expired"
        case .forbidden:
            return "Access denied"
        case .invalidURL:
            return "Invalid URL"
        case .invalidResponse:
            return "Invalid response"
        case .rateLimited:
            return "Too many requests"
        case .serverError(let code):
            return "Server error (\(code))"
        case .unexpectedStatus(let code):
            return "Unexpected error (\(code))"
        case .decodingFailed(let detail):
            return "Data error: \(detail)"
        }
    }
}

File: android/app/src/main/java/com/mysdamanager/app/widget/TaskService.kt
package com.mysdamanager.app.widget

import android.content.Context
import android.util.Log
import org.json.JSONArray
import org.json.JSONObject
import java.io.BufferedReader
import java.io.InputStreamReader
import java.io.OutputStreamWriter
import java.net.HttpURLConnection
import java.net.URL

/**
 * HTTP client for the MySDAManager widget API.
 * Handles fetching tasks and completing tasks via the REST API.
 *
 * Uses HttpURLConnection (no external dependencies) to keep the widget lightweight.
 * All methods are synchronous and should be called from a background thread/coroutine.
 */
class TaskService(private val context: Context) {

    companion object {
        private const val TAG = "TaskService"
        private const val BASE_URL = "https://mysdamanager.com/api/v1/tasks/widget"
        private const val REFRESH_URL = "https://mysdamanager.com/api/v1/auth/refresh"
        private const val CONNECT_TIMEOUT = 15_000 // 15 seconds
        private const val READ_TIMEOUT = 15_000    // 15 seconds
        private const val USER_AGENT = "MySDAManager-Android-Widget/1.0"
    }

    private val tokenManager = TokenManager(context)

    // MARK: - Fetch Tasks

    /**
     * Fetches the user's pending tasks from the API.
     * Returns a WidgetResponse on success, or null on failure.
     * Falls back to cached data if the network request fails.
     *
     * @param limit Maximum number of tasks to fetch (default 10)
     */
    fun fetchTasks(limit: Int = 10): TaskFetchResult {
        val token = tokenManager.getSessionToken()
            ?: return TaskFetchResult.NotAuthenticated

        return try {
            val url = URL("$BASE_URL?limit=$limit")
            val connection = (url.openConnection() as HttpURLConnection).apply {
                requestMethod = "GET"
                setRequestProperty("Authorization", "Bearer $token")
                setRequestProperty("Accept", "application/json")
                setRequestProperty("User-Agent", USER_AGENT)
                connectTimeout = CONNECT_TIMEOUT
                readTimeout = READ_TIMEOUT
            }

            try {
                val responseCode = connection.responseCode

                when (responseCode) {
                    200 -> {
                        val responseBody = readStream(connection)
                        val json = JSONObject(responseBody)
                        val response = WidgetResponse.fromJson(json)

                        // Cache successful response
                        tokenManager.cacheTasks(json.getJSONArray("data").toString())

                        TaskFetchResult.Success(response.data, response.count)
                    }

                    401 -> {
                        // Try token refresh
                        if (refreshToken()) {
                            // Retry with new token
                            fetchTasks(limit)
                        } else {
                            tokenManager.clearTokens()
                            TaskFetchResult.SessionExpired
                        }
                    }

                    403 -> TaskFetchResult.Forbidden
                    429 -> fallbackToCacheOrError("Rate limited")
                    in 500..599 -> fallbackToCacheOrError("Server error ($responseCode)")
                    else -> fallbackToCacheOrError("Unexpected status: $responseCode")
                }
            } finally {
                connection.disconnect()
            }
        } catch (e: Exception) {
            Log.e(TAG, "Failed to fetch tasks", e)
            fallbackToCacheOrError(e.message ?: "Network error")
        }
    }

    // MARK: - Complete Task

    /**
     * Marks a task as completed via the API.
     *
     * @param taskId The ID of the task to complete
     * @return true if the task was successfully completed
     */
    fun completeTask(taskId: String): Boolean {
        val token = tokenManager.getSessionToken() ?: return false

        return try {
            val url = URL(BASE_URL)
            val connection = (url.openConnection() as HttpURLConnection).apply {
                requestMethod = "POST"
                setRequestProperty("Authorization", "Bearer $token")
                setRequestProperty("Content-Type", "application/json")
                setRequestProperty("User-Agent", USER_AGENT)
                connectTimeout = CONNECT_TIMEOUT
                readTimeout = READ_TIMEOUT
                doOutput = true
            }

            try {
                val body = JSONObject().apply {
                    put("taskId", taskId)
                }

                OutputStreamWriter(connection.outputStream).use { writer ->
                    writer.write(body.toString())
                    writer.flush()
                }

                connection.responseCode == 200
            } finally {
                connection.disconnect()
            }
        } catch (e: Exception) {
            Log.e(TAG, "Failed to complete task: $taskId", e)
            false
        }
    }

    // MARK: - Token Refresh

    /**
     * Attempts to refresh the session token using the stored refresh token.
     * @return true if the token was successfully refreshed
     */
    private fun refreshToken(): Boolean {
        val refreshToken = tokenManager.getRefreshToken() ?: return false

        return try {
            val url = URL(REFRESH_URL)
            val connection = (url.openConnection() as HttpURLConnection).apply {
                requestMethod = "POST"
                setRequestProperty("Content-Type", "application/json")
                setRequestProperty("User-Agent", USER_AGENT)
                connectTimeout = CONNECT_TIMEOUT
                readTimeout = READ_TIMEOUT
                doOutput = true
            }

            try {
                val body = JSONObject().apply {
                    put("refreshToken", refreshToken)
                }

                OutputStreamWriter(connection.outputStream).use { writer ->
                    writer.write(body.toString())
                    writer.flush()
                }

                if (connection.responseCode == 200) {
                    val responseBody = readStream(connection)
                    val json = JSONObject(responseBody)

                    val newSessionToken = json.getString("sessionToken")
                    tokenManager.storeSessionToken(newSessionToken)

                    if (json.has("refreshToken")) {
                        tokenManager.storeRefreshToken(json.getString("refreshToken"))
                    }

                    true
                } else {
                    false
                }
            } finally {
                connection.disconnect()
            }
        } catch (e: Exception) {
            Log.e(TAG, "Failed to refresh token", e)
            false
        }
    }

    // MARK: - Cache Fallback

    /**
     * Attempts to return cached tasks when the network request fails.
     * Returns an error result if no cache is available.
     */
    private fun fallbackToCacheOrError(errorMessage: String): TaskFetchResult {
        val cachedJson = tokenManager.getCachedTasks()
        if (cachedJson != null) {
            try {
                val jsonArray = JSONArray(cachedJson)
                val tasks = TaskItem.fromJsonArray(jsonArray)
                if (tasks.isNotEmpty()) {
                    return TaskFetchResult.Success(tasks, tasks.size, isCached = true)
                }
            } catch (e: Exception) {
                Log.e(TAG, "Failed to parse cached tasks", e)
            }
        }
        return TaskFetchResult.Error(errorMessage)
    }

    // MARK: - Helpers

    /**
     * Reads the response body from an HttpURLConnection.
     */
    private fun readStream(connection: HttpURLConnection): String {
        val reader = BufferedReader(InputStreamReader(connection.inputStream))
        val response = StringBuilder()
        var line: String?
        while (reader.readLine().also { line = it } != null) {
            response.append(line)
        }
        reader.close()
        return response.toString()
    }
}

/**
 * Sealed class representing the result of a task fetch operation.
 */
sealed class TaskFetchResult {
    /**
     * Tasks fetched successfully.
     * @param tasks List of task items
     * @param totalCount Total number of tasks available
     * @param isCached Whether the data came from the local cache
     */
    data class Success(
        val tasks: List<TaskItem>,
        val totalCount: Int,
        val isCached: Boolean = false
    ) : TaskFetchResult()

    /** User is not authenticated (no token stored). */
    data object NotAuthenticated : TaskFetchResult()

    /** Session token has expired and could not be refreshed. */
    data object SessionExpired : TaskFetchResult()

    /** Access denied (wrong permissions). */
    data object Forbidden : TaskFetchResult()

    /** An error occurred during the fetch. */
    data class Error(val message: String) : TaskFetchResult()
}
